### PR TITLE
SA-258 add table diffs

### LINKED ISSUE #169: Support Diffs for tables
pass table and snapshot id

get a dif of tables from 2 snapshots

### PR DESCRIPTION
**Pull request by cordeezy2**

### CODE REVIEW

**Comment by cordeezy2** [comment]
@justinjeffery-ipf Thoughts on this?
any recommendations ?

**Comment by ghost** [comment]
I would think this would just be a function in the Table class: https://github.com/community-fabric/python-ipfabric/blob/00fdaf3f113b0ecb8daab27867db95db9fa0d360/ipfabric/models.py#L11

This would be similar to how Intents work: https://github.com/community-fabric/python-ipfabric/blob/00fdaf3f113b0ecb8daab27867db95db9fa0d360/ipfabric/intent.py#L103

**Comment by ghost** [comment]
Default to ignore certain columns:

- id
- _id
- rd (Routing Domain)
- stpDomain

User must be able to specify which columns they want to compare (and/or columns to ignore).  For instance Inventory > Devices ignore `uptime` and `memoryUtilization`.

Using deepdiff:
https://zepworks.com/deepdiff/current/exclude_paths.html#exclude-paths-label

File: ipfabric/tools/table_diff.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
+from __future__ import annotations
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from ipfabric import IPFClient
+    from ipfabric.snapshot_models import Snapshot
+import logging
+from typing import Any
+import jsondiff as jd
```
Comment (ghost) [inline ipfabric/tools/table_diff.py]: Use deepdiff as it has more support and a lot more features we could possibly use.
Also make sure to do `poetry add deepdiff && poetry update` and commit toml and lock file.

Used in other project already: https://github.com/community-fabric/python-copy-ipfabric-settings/blob/3d0847a25631f763e46ddd627ccdabdffde42abd/ipf_settings.py#L41

**Comment by ghost** [comment]
Investigate deepdiff optimization: https://zepworks.com/deepdiff/current/optimizations.html

Example benchmark: https://github.com/seperman/deepdiff-benchmark

Probably use Routing Table for testing?

**Comment by cordeezy2** [comment]
@justinjeffery-ipf 
heres the hashing idea i discussed this morning, let me know what you think. 
I like this method, as it doesnt introduce any additional package requirements and still provides the basic functionality requested.

TODO:
add some benchmarks for performance of diffs

**Comment by ghost** [comment]
@cordeezy2 merge develop into branch so I could check changes

File: ipfabric/models.py [hunk 2]
```diff
@@ -104,6 +105,94 @@ def count(
...
+            ):
+        if columns_ignore is None:
+            columns_ignore = ['id']
+        if 'id' not in columns_ignore:
+            columns_ignore.append('id')
+        # get all columns for the table
+        table_cols = self.client._get_columns(self.endpoint)
+
+        # determine which columns to use in query
+        if columns or columns_ignore:
```
Comment (ghost) [inline ipfabric/models.py]: This will always be True as you are setting `columns_ignore` above.

File: ipfabric/models.py [hunk 3]
```diff
@@ -104,6 +105,94 @@ def count(
             snapshot=self.snapshot,
         )
 
+    def _compare_determine_columns(self, table_columns, columns, columns_ignore): #NOSONAR
+        cols_for_return = []
+        # user does not pass any columns
+        if columns is None and columns_ignore is None:
```
Comment (ghost) [inline ipfabric/models.py]: See note below, `columns_ignore` is set to ['id'] so this if will never execute.

File: ipfabric/tools/table_diff.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                    valid_tables = [method for method in dir(return_client) if method not in TECH_TABLE_METHODS_IGNORE]
+                    logger.error(f"unable to determine nested table {tech_table}")
+                    logger.info(f"valid tables: \n {valid_tables}")
+                return return_client
+        else:
+            logger.error(f"unable to determine nested table {tech_table}")
+            logger.info(f"valid tables: \n {valid_tables}")
+            logger.info(f"{HELP_STR}")
+
+    def _check_inv_table_attr(self, inv_table: str):
```
Comment (ghost) [inline ipfabric/tools/table_diff.py]: Is this still needed?  Don't see it referenced anywhere

**Review by ghost — CHANGES_REQUESTED** [review]

File: ipfabric/tools/table_diff.py (hunk 4 above)
Comment (cordeezy2) [reply ipfabric/tools/table_diff.py]: I removed this file, sorry for the confusion, my bad :pensive:

File: ipfabric/models.py (hunk 3 above)
Comment (cordeezy2) [reply ipfabric/models.py]: 2250f9f corrected in latest commits

File: ipfabric/models.py (hunk 2 above)
Comment (cordeezy2) [reply ipfabric/models.py]: corrected in latest commits 2250f9f

File: ipfabric/tools/table_diff.py (hunk 1 above)
Comment (cordeezy2) [reply ipfabric/tools/table_diff.py]: removed file

**Comment by cordeezy2** [comment]
@justinjeffery-ipf should be ready for merge, let me know if theres anything else missing

**Comment by cordeezy2** [comment]
all Test are passing in my local env, odd

**Comment by cordeezy2** [comment]
i was testing on 3.8.13, CI is running on 3.8.14

**Pull request merged**